<?xml version="1.0"?>

<launch>
    <!-- Experiment settings -->
    <arg name="sim" default="false"/>
    <arg name="gazebo_gui" default="false"/>
    <arg name="record" default="true"/>
    <arg name="drone_name" default="hovergames"/>
    <arg name="run_px4_control_interface" default="true"/>
    <arg name="px4_control_interface_param_file" default="$(find px4_tools)/config/px4_control_interface_default.yaml"/>

    <!-- Controller settings -->
    <arg name="controller_param_file" default="$(find drone_toolbox_ext_control_template)/config/controller.yaml"/>

    <!-- Vicon settings -->
    <arg name="publish_tf" default="true"/>
    <arg name="reset_z_axis" default="true"/>
    <arg name="only_use_object_specific" default="true"/>
    <arg name="object_names" default="[drone1]"/>
    <arg name="object_msg_types" default="[geometry_msgs/PoseStamped]"/>
    <arg name="object_frame_ids" default="[map]"/>
    <arg name="object_publish_topics" default="[/mavros/vision_pose/pose]"/>
    <arg name="object_frequency_divider" default="[1]"/>

    <!-- Load robot_description and publish the states on tf -->
    <include file="$(find drone_description)/launch/drone_description.launch">
        <arg name="drone_name" value="$(arg drone_name)"/>
    </include>

    <!-- Launch Gazebo if sim is set to true -->
    <group if="$(arg sim)">
        <include file="$(find drone_gazebo)/launch/gazebo_world.launch">
            <arg name="gui" value="$(arg gazebo_gui)"/>
        </include>
    </group>

    <!-- Launch PX4 control interface, MAVROS, PX4 SITL and spawn vehicle -->
    <include file="$(find px4_tools)/launch/px4_tools.launch">
        <arg name="sim" value="$(arg sim)"/>
        <arg name="drone_name" value="$(arg drone_name)"/>
        <arg name="run_px4_control_interface" value="$(arg run_px4_control_interface)"/>
        <arg name="px4_control_interface_param_file" value="$(arg px4_control_interface_param_file)"/>
    </include>

    <!-- Launch controller -->
    <node name="controller" pkg="drone_toolbox_ext_control_template" type="controller_node" required="true" output="screen">
        <param name="is_sim" value="$(arg sim)"/>
        <rosparam command="load" file="$(arg controller_param_file)"/>
    </node>

    <!-- Print warning message for lab experiments -->
    <group unless="$(arg sim)">
        <arg name="msg_warn" default="Usage for lab experiments is at your own risk! This is a template controller giving example commands that can potentially be dangerous in a real-world scenario. Use this template controller package as a basis to create your own controller."/>
        <node name="pub_warn" pkg="rostopic" type="rostopic" args="pub /msg_warn std_msgs/String '$(arg msg_warn)'" output="screen"/>
        <node name="print_warn" pkg="rostopic" type="rostopic" args="echo /msg_warn" output="screen"/>
    </group>

    <!-- Launch vicon bridge if sim is set to false -->
    <group unless="$(arg sim)">
        <include file="$(find vicon_bridge)/launch/vicon.launch">
            <arg name="publish_tf" value="$(arg publish_tf)"/>
            <arg name="reset_z_axis" value="$(arg reset_z_axis)"/>
            <arg name="only_use_object_specific" value="$(arg only_use_object_specific)"/>
            <arg name="object_names" value="$(arg object_names)"/>
            <arg name="object_msg_types" value="$(arg object_msg_types)"/>
            <arg name="object_frame_ids" value="$(arg object_frame_ids)"/>
            <arg name="object_publish_topics" value="$(arg object_publish_topics)"/>
            <arg name="object_frequency_divider" value="$(arg object_frequency_divider)"/>
        </include>
    </group>

    <!-- Launch ROS dynamic reconfigure -->
    <group if="$(arg sim)">
        <node name="rqt_reconfig" pkg="rqt_reconfigure" type="rqt_reconfigure"/>
    </group>

    <!-- Launch pre-configured RViz -->
    <group if="$(arg sim)">
        <node name="rviz" pkg="rviz" type="rviz" args="-d $(find drone_toolbox_ext_control_template)/rviz/template_controller.rviz"/>
    </group>

    <!-- Record a ROS bag -->
    <group if="$(arg record)">
        <node pkg="rosbag" name="record" type="record" args="-o template_controller
        /mavros/setpoint_raw/local
        /mavros/setpoint_raw/attitude
        /mavros/setpoint_raw/roll_pitch_yawrate_thrust
        /mavros/local_position/odom"
        />
    </group>
</launch>
